components/script/dom/bluetoothpermissionresult.rs

```diff
@@ -0,0 +1,219 @@
+            PermissionDescriptorType::Bluetooth(ref d) => d,
+            _ => return Err(Error::Type(DESCRIPTOR_TYPE_ERROR.to_owned())),
+        };
+
+        // Step 1.
+        if (options.filters.is_some() && options.acceptAllDevices) ||
+           (options.filters.is_none() && !options.acceptAllDevices) {
+            return Err(Error::Type(OPTIONS_ERROR.to_owned()));
+        }
+
+        // Step 2.
+        let sender = response_async(promise, self);
+        let bluetooth = self.get_bluetooth();
+        Ok(bluetooth.request_bluetooth_devices(promise, &options.filters, &options.optionalServices, sender))
+        // NOTE: Step 3 is in `handle_response` function.
+    }
+}
+
+
+// TODO: Make a BluetoothResponse variant which returns with a vector of devices that matches.
+impl AsyncBluetoothListener for BluetoothPermissionResult {
+    fn handle_response(&self, response: BluetoothResponse, promise_cx: *mut JSContext, promise: &Rc<Promise>) {
+        match response {
+            BluetoothResponse::RequestDevice(device) => {
+                let bluetooth = &self.get_bluetooth();
+                let mut device_instance_map = bluetooth.get_device_map().borrow_mut();
+                if let Some(existing_device) = device_instance_map.get(&device.id.clone()) {
+                    // https://webbluetoothcg.github.io/web-bluetooth/#bluetoothpermissionresult
+                    // Step 3.
+                    *self.devices.borrow_mut() = vec!(JS::from_ref(&**existing_device));
+                    // https://w3c.github.io/permissions/#dom-permissions-request
+                    // Step 8.
+                    return promise.resolve_native(promise_cx, &self.parent);
+                }
+                let bt_device = BluetoothDevice::new(&self.global(),
+                                                     DOMString::from(device.id.clone()),
+                                                     device.name.map(DOMString::from),
+                                                     bluetooth);
+                device_instance_map.insert(device.id, JS::from_ref(&bt_device));
+                // https://webbluetoothcg.github.io/web-bluetooth/#bluetoothpermissionresult
+                // Step 3.
+                *self.devices.borrow_mut() = vec!(JS::from_ref(&bt_device));
+                // https://w3c.github.io/permissions/#dom-permissions-request
+                // Step 8.
+                promise.resolve_native(promise_cx, &self.parent);
+            },
+            _ => promise.reject_error(promise_cx, Error::Type("Something went wrong...".to_owned())),
+        }
+    }
+}
+
+impl BluetoothPermissionResultMethods for BluetoothPermissionResult {
+    // https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothpermissionresult-devices
+    fn Devices(&self) -> Vec<Root<BluetoothDevice>> {
+        let device_vec: Vec<Root<BluetoothDevice>> =
+            self.devices.borrow().iter().map(|d| Root::from_ref(&**d)).collect();
```

Comment:
Yes It's called `deref()` which belongs to the `Deref` trait, and `*` is a shorthand for that.